File: packages/antgrid-wire/src/relay-slot.ts
/**
 * The app's relay slot id: the `hello.deviceId` one app socket presents,
 * `<accountDeviceUuid>#<machineDeviceUuid>`.
 *
 * The relay arbitrates per `hello.deviceId` and supersedes an equal epoch, so
 * an app holding several machines open at once needs one slot per machine.
 * A slot identifies the app's machine-scoped control connection and native
 * route. Everything keyed by the ACCOUNT device — the bridge's
 * `paired-phones.ts` lookups and central revocation lookup — goes through
 * `baseSlotDeviceId` first.
 *
 * Stripping the scope cannot admit anyone: the native payload path
 * authorizes by endpoint ID from the authorization snapshot
 * (`acceptPeer`, `bridge/src/peer/native-host-connection.ts`), which a
 * stripped or forged slot never resolves to.
 *
 * Hand-mirrored by `packages/antgrid_relay_client/lib/src/relay_slot.dart`
 * (which also mints slots — only the app ever does) — keep the two in
 * lockstep. A divergence is an admission failure, not a type error.
 */
const SLOT_SEPARATOR = "#";

/** Build the app's relay slot id for one (account device, machine) pair. The
 *  one TS copy of the separator — mirrored by
 *  `packages/antgrid_relay_client/lib/src/relay_slot.dart`'s `relaySlotId`,
 *  which also mints slots (only the app ever does). Keep the two in lockstep. */
export function relaySlotId(deviceUuid: string, machineDeviceId: string): string {
  return `${deviceUuid}${SLOT_SEPARATOR}${machineDeviceId}`;
}

/** The bare account `deviceUuid` behind a route id. Unscoped ids pass through,
 *  so a client that dials without a slot keeps working. */
export function baseSlotDeviceId(routeId: string): string {
  const i = routeId.indexOf(SLOT_SEPARATOR);
  return i < 0 ? routeId : routeId.slice(0, i);
}

/** True when `routeId` is a slot scoped under account device `deviceId`.
 *  False for `deviceId` itself — callers that want both check equality too. */
export function isSlotOf(routeId: string, deviceId: string): boolean {
  return routeId.startsWith(deviceId + SLOT_SEPARATOR);
}

File: packages/antgrid_relay_client/lib/src/relay_slot.dart
/// The app's relay slot id: the `hello.deviceId` one app socket presents.
///
/// The relay arbitrates per `hello.deviceId` and rejects an epoch that is lower
/// OR EQUAL with `SUPERSEDED`, so every socket the app holds at once needs its
/// own slot. Scoping by the machine each socket serves gives exactly that: one
/// slot per (app device, machine) pair, and the app's own sockets never
/// arbitrate against each other.
///
/// The scope is the AGENT's deviceUuid, so a bridge receiving a slot can tell
/// whether it was addressed at itself — see the presence guard in
/// `bridge/src/relay-client.ts`.
///
/// This is only ever a TRANSPORT address. The E2E transcript binds the bare
/// account `deviceUuid` on both sides, so the account identity a handshake
/// proves is unaffected by how the socket is addressed.
///
/// The relay's `DEVICE_ID` schema (`packages/antgrid-wire/src/relay-protocol.ts`)
/// must admit the separator, in `hello.deviceId` AND in a route header's `to` —
/// a schema miss is `PROTOCOL_VIOLATION` before the signature is looked at.
///
/// Mirrored by `packages/antgrid-wire/src/relay-slot.ts` (the one TS copy, used
/// by both bridge and relay) — keep the two in lockstep.
String relaySlotId(String deviceUuid, String machineDeviceId) =>
    '$deviceUuid$kRelaySlotSeparator$machineDeviceId';

/// Reduce a relay slot id to the bare account `deviceUuid` that identity
/// lookups and the E2E transcript are keyed by. Returns the input unchanged
/// when it carries no scope.
String baseSlotDeviceId(String slotId) {
  final i = slotId.indexOf(kRelaySlotSeparator);
  return i < 0 ? slotId : slotId.substring(0, i);
}

/// `#`, not `.`: a dot is the retired compound `<deviceUuid>.<projectId>` form,
/// and reusing it would make a slot indistinguishable from one of those.
const String kRelaySlotSeparator = '#';
